src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -62,19 +77,396 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
     }
   }
 
-  private Future<Void> savePermList(List<Perm> permList, Context vertxContext, String tenantId) {
+  private Future<List<Permission>> getPermsForModule(ModuleId moduleId, List<OkapiPermission> permSet,
+      Context vertxContext, String tenantId) {
+    Promise<List<Permission>> promise = Promise.promise();
+    
+    try {
+      getPermsByModule(moduleId, vertxContext, tenantId)
+        .onComplete(getResult -> {
+          if (getResult.failed()) {
+            Throwable cause = getResult.cause();
+            logger.error(cause.getMessage(), cause);
+            promise.fail(cause);
+            return;
+          }
+
+          List<Permission> existing = getResult.result();
+          if (existing.isEmpty() && permSet != null) {
```

Comment:
This should be rewritten to make use of Future compose: https://github.com/folio-org/raml-module-builder/blob/master/doc/futurisation.md#result-and-error-handling
This will avoid most error handling code and the try-catch-clause and the promise.

    return getPermsByModule(moduleId, vertxContext, tenantId)
      .compose(existing -> {
        if (! existing.isEmpty() || permSet == null) {
          return Future.succeededFuture(existing);
        }
        ...
        return CompositeFuture.all(futures).map(ret);
      })
      .onFailure(e -> logger.error(e.getMessage(), e));

Reply:
I agree, much better; thanks.